bmap-rs/src/main.rs

```diff
@@ -87,12 +94,32 @@ fn setup_input(path: &Path) -> Result<Decoder> {
     }
 }
 
-fn copy(c: Copy) -> Result<()> {
-    if !c.image.exists() {
+async fn setup_remote_input(path: &Path) -> Result<Decoder> {
+    if path.extension().unwrap() != "gz" {
+        bail!("Image file format not implemented")
+    }
+    let url = Url::parse(path.to_str().unwrap())?;
+    let data = reqwest::get(url).await?.bytes().await?;
```

Comment:
does this mean we await for all of the data to be downloaded to memory first ?

Reply:
Yes, as the execution is kind of "lazy", in order for the compiler to consider that `reqwest::get(url)` is a value where we can call `.bytes()` we need to tell him to check whether it's value is ready, and wait if not.

Reply:
that will work quite badly as your image is likely many gigs; You'll want to read the bytes streaming to do that from request to a futures AsyncRead you can do something like:

```rust
    let r = reqwest::get("http://collabora.com").await.unwrap();
    let mut b = r
        .bytes_stream()
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))
        .into_async_read();
```

Reply:
but ofcourse this means the whole pipeline needs to be async; The other option is to have a buffer implementation that uses an async reader on one side and provide the sync interfaces on another.. But that's probably not worth it

Reply:
I've tried using `.into_async_read()` but I can't get it it's output to be compatible with the `GzDecoder`. I think one of them is `AsyncBufRead` and the other one is `AsyncRead`. So the way I've found to channel `bytes_stream` into the `GzDecoder` is with `StreamReader`

Reply:
flate2::GzDecoder only works with sync Read or BufRead (depending on the variant); Also mind that both the futures crate and tokio define Async{Buf,}Read traits, so if you use `into_async_read` as in my example the consumers need to use the Futures trait while if you use e.g. `StreamReader` your consumers should use the Tokio traits.

Reply:
I see, I just checked and we could change the use of Tokio traits to futures and not use `StreamReader`. But we are using Tokio traits for `AsyncDiscarder`, if we change to futures it requires the implementation of `poll_read_vectored` in addition to the `poll_read` we have. I'm not sure if that's worth it. Since it seems that the problem is solved with Tokio traits and `StreamReader`

Reply:
I've added a commit on top with the changes in order to use `futures` traits. With the idea of comparing and squashing it if we decide we want those changes.